### DISCUSSION TITLE
New GPU strategy?

### DISCUSSION THREAD

**Discussion by sonora** [osmandapp/OsmAnd#16306] [Ideas]
- I count about OsmAnd 300 issues we have collected so far which are related to us introducing OpenGL, with more than 50 still open.
- OpenGL development itself had been discontinued in 2017.
- Some platforms have already discontinued supporting it  (notably Apple. Also Google has initiatives focusing on Vulcan.).
- All successor interfaces, most notably Vulcan, are not downward compatible with OpenGL, which paves any way forward for us with even more effort and issues around backward compatibility to interim interfaces we had decided to support.

I guess we should start some deliberation if using OpenGL (even plans to make it our default?!) is really the right current strategy to offload our map rendering to the GPU?

**Comment by vshcherb** [comment]
Indeed, let me clarify some things as separate statements so probably it will be clear the logic behind it.
- There are 2 possible ways to render on device as Bitmap (Skia - Chrome, Android / Cocoa - iOS) as Triangles -> Shaders (OpenGL, Vulkan, Metal).
- OsmAnd OpenGL uses combined approach renders Symbols (Path, Icons, Texts) as Triangles and Roads / Maps as Bitmaps. It works on PC, Android, iOS and all sorts of combinations
- OsmAnd pre-OpenGL only works on **Android** and renders everything on **Bitmap**. It's no go strategy cause doesn't allow to have 3D, doesn't utilize any GPU power from device, though with this approach it's easier to render things on canvas as it has more methods.
-  OpenGL, Vulkan, Metal - it's the list of technologies over Rendering Pipeline which is defined by GPU, Vulkan provides a higher level abstraction over OpenGL but in future will work without OpenGL, Metal only works on Apple devices. In the end it doesn't matter cause Rendering Pipeline is going to stay on devices. **Rendering Pipeline**:
   - Consists of points, triangles, textures
   - Has fragmentation, color shaders  etc
   - Works with 3D geometry (instead of 2D on Canvas)
   - Requires elaborate resource manager of textures 
- OsmAnd code works with Rendering Pipeline on iOS since 2015, on Android it was a hard switch last month and with much much bigger variety of devices and use cases requires 3-6 months to stabilize. 
- OsmAnd code works with Rendering Pipeline only through OpenGL though it requires 1-2 month of work to replace to Metal or Vulkan. OpenGL is supported and going to be supported on Android for a long time, Apple is not clear. Vulkan is not supported even on most of devices produced in 2022, so it's going to take 5 years to start dominating.

In a nutshell, there is a rendering pipeline and there is a simple canvas drawing approach. Canvas is simple to start and stable but it doesn't utlitize GPU, doesn't scale performance and doesn't provide 3D. That's pertty much it

**Reply by sonora** [reply]
Thanks for taking the time to comment! 

And yes, regarding "Altenative impementations", Wikipedia maybe summarizes it best by stating on one side

> Apple deprecated OpenGL in iOS 12 and macOS 10.14 Mojave in favor of Metal, but it is still available as of macOS 13 Ventura (including Apple silicon devices).


as well as saying:

> There are several projects which attempt to implement OpenGL on top of Vulkan. The Vulkan backend for Google's ANGLE achieved OpenGL ES 3.1 conformance in July 2020.
